notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb

```diff
@@ -0,0 +1,1070 @@
+    "        xpos = (plt_num%40) % cols\n",
+    "        ypos = ((plt_num%40) // cols) # // to make it an int.\n",
+    "\n",
+    "        # make the subplot\n",
+    "        ax = plt.subplot2grid((rows, cols), (ypos, xpos))\n",
+    "    \n",
+    "        if 'i2d' in img:\n",
+    "            cat = Table.read(img.replace('i2d.fits', cat_suffix))\n",
+    "            cmap = 'gist_gray'\n",
+    "        else:\n",
+    "            cmap = 'tab20c_r'\n",
+    "            \n",
+    "        # plot the image\n",
+    "        with fits.open(img) as hdu:\n",
+    "            ax.imshow(hdu[1].data, vmin=0, vmax=0.3, origin='lower', cmap=cmap)\n",
+    "            title = f\"obs{hdu[0].header['OBSERVTN']} {hdu[0].header['PUPIL']}\"\n",
+    "    \n",
+    "        # also plot the associated catalog\n",
+    "        extended_sources = cat[cat['is_extended'] == True]\n",
+    "        point_sources = cat[cat['is_extended'] == False]\n",
+    "        \n",
+    "        for color, sources in zip(['darkred', 'black'], [extended_sources, point_sources]):\n",
+    "            # plotting the sources\n",
+    "            ax.scatter(sources['xcentroid'], sources['ycentroid'], s=20, facecolors='None', edgecolors=color, alpha=0.9)\n",
+    "    \n",
+    "            # adding source labels \n",
+    "            for i, source_num in enumerate(sources['label']):\n",
+    "                ax.annotate(source_num, \n",
+    "                            (sources['xcentroid'][i]+0.5, sources['ycentroid'][i]+0.5), \n",
+    "                            fontsize=8,\n",
+    "                            color=color)\n",
+    "        if 'i2d' in img:\n",
+    "            ax.set_title(f\"{title} combined image\")\n",
+    "        else:\n",
+    "            ax.set_title(f\"{title} segmentation map\")\n",
+    "        \n",
+    "        # zooming in on a smaller region\n",
+    "        ax.set_xlim(xmin, xmax)\n",
+    "        ax.set_ylim(ymin, ymax)\n",
+    "    \n",
+    "    # Helps to make the axes not overlap ; you can also set this manually if this doesn't work\n",
+    "    plt.tight_layout()\n",
+    "\n",
+    "    return fig"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "0599d52e-7a92-4c16-92e7-59ae62cfe8d9",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "default_fig = plot_image_and_segmentation_map(image3_i2d, image3_segm)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d1addba0-ae56-4565-b12a-3ded03d9b0bf",
+   "metadata": {},
+   "source": [
+    "#### ImViz\n",
+    "\n",
+    "Similarly to DS9, ImViz allows you to interactively view these images and the corresponding source catalog as well."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "7bec8a55",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "imviz = Imviz()\n",
+    "viewer = imviz.default_viewer\n",
+    "\n",
+    "for img in image3_i2d:\n",
```

Comment:
I suggest splitting this into 2 for loops, one for the images and one for the catalogs. The linking can happen between the two loops only once.

Reply:
@camipacifici can you expand on why you would do this? It seems inefficient to do it that way since I have to get the catalog name from the i2d, but perhaps there's an imviz reason to do this?

Reply:
I think I answered my own question trying to apply your next comment. There's a linking issue if you use sky coordinates and the markers that doesn't exist when you are using x/y pixels